ath79: add support for D-Link COVR-P2500 A1

jokujossai opened the pull request:
Specifications:
* QCA9563, 16 MiB flash, 128 MiB RAM, 2T2R 802.11n
* QCA9886 2T2R 801.11ac Wave 2
* QCA7550 Homeplug AV2 1300
* AR8337, 3 Gigabit ports (1, 2: LAN; 3: WAN)

To make use of PLC functionality, firmware needs to be provided via plchost (QCA7550 comes without SPI NOR), patched with the Network Password and MAC.

Flashing via OEM Web Interface
* Flash 'factory.bin' using web-interface
* Wait until firmware succesfully installed and device booted
* Hold down reset button to reset factory defaults (~10 seconds)

Flashing via Recovery Web Interface:
* Hold down reset button during power-on (~10 seconds)
* Recovery Web UI is at 192.168.0.50, no DHCP.
* Flash 'recovery.bin' with scripts/flashing/dlink_recovery_upload.py (Recovery Web UI does not work with modern OSes)

Return to stock
* Hold down reset button during power-on (~10 seconds)
* Recovery Web UI is at 192.168.0.50, no DHCP.
* Flash unencrypted stock firmware with scripts/flashing/dlink_recovery_upload.py (Recovery Web UI does not work with modern OSes)

Co-developed-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Daniel Linjama <daniel@dev.linjama.com>

jokujossai committed:
ath79: add support for D-Link COVR-P2500 A1

Specifications:
* QCA9563, 16 MiB flash, 128 MiB RAM, 2T2R 802.11n
* QCA9886 2T2R 801.11ac Wave 2
* QCA7550 Homeplug AV2 1300
* AR8337, 3 Gigabit ports (1, 2: LAN; 3: WAN)

To make use of PLC functionality, firmware needs to be
provided via plchost (QCA7550 comes without SPI NOR),
patched with the Network Password and MAC.

Flashing via OEM Web Interface
* Flash 'factory.bin' using web-interface
* Wait until firmware succesfully installed and device booted
* Hold down reset button to reset factory defaults (~10 seconds)

Flashing via Recovery Web Interface:
* Hold down reset button during power-on (~10 seconds)
* Recovery Web UI is at 192.168.0.50, no DHCP.
* Flash 'recovery.bin' with
  scripts/flashing/dlink_recovery_upload.py
  (Recovery Web UI does not work with modern OSes)

Return to stock
* Hold down reset button during power-on (~10 seconds)
* Recovery Web UI is at 192.168.0.50, no DHCP.
* Flash unencrypted stock firmware with
  scripts/flashing/dlink_recovery_upload.py
  (Recovery Web UI does not work with modern OSes)

Co-developed-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Daniel Linjama <daniel@dev.linjama.com>

jokujossai commented:
Device support files are originally implemented by @s-2 but firmware-utils does not support encrypted factory image. Images built with Image Builder are availbale on [openwrt_covr_p2500 repository](https://github.com/jokujossai/openwrt_covr_p2500) (also encrypted images for OEM firmware upgrade).

I wrote dlink_recovery_upload.py script that allows flashing unencrypted images using recovery web UI on modern Linux and Windows computers so encrypted factory image is no longer needed for installation.

s-2 commented:
Awesome, thanks for your work on this device!

I wonder if we should add the default mappings for the LEDs (it seems I totally forgot about that when working with this device back then), this can be done for wireless interfaces in the .dts, e.g.
```
linux,default-trigger = "phy0tpt";
```
or `phy1tpt` respectively; for LAN this should probably go to `base-files/etc/board.d/01_leds`,
I think `ucidef_set_led_netdev` for the `eth0` node would do (not sure what's the difference with `ucidef_set_led_switch`, probably for selecting individual ports from the switch? But we just have one LED for all 3 ports here, so I guess it should not be limited to either WAN nor LAN).

jokujossai commented:
> Awesome, thanks for your work on this device!
> 
> I wonder if we should add the default mappings for the LEDs (it seems I totally forgot about that when working with this device back then), this can be done for wireless interfaces in the .dts, e.g.
> 
> ```
> linux,default-trigger = "phy0tpt";
> ```
> 
> or `phy1tpt` respectively; for LAN this should probably go to `base-files/etc/board.d/01_leds`, I think `ucidef_set_led_netdev` for the `eth0` node would do (not sure what's the difference with `ucidef_set_led_switch`, probably for selecting individual ports from the switch? But we just have one LED for all 3 ports here, so I guess it should not be limited to either WAN nor LAN).

I added `phy0radio` trigger for the `green:wlan5g` LED and `phy1radio` trigger for the `green:wlan2g` LED. For the `green:lan` LED I added:
```
ucidef_set_led_switch "lan" "LAN" "green:lan" "switch0" "0x0e" "" "link"
```
The port mask `0x0e` selects ports 1, 2 (LAN) and 3 (WAN). The functionality of the LEDs is now similar to the stock firmware.

hauke commented on scripts/flashing/dlink_recovery_upload.py:
```diff
@@ -0,0 +1,337 @@
+
+    # Parse tcp_info structure from /usr/include/netinet/tcp.h
+    assert os.path.isfile("/usr/include/netinet/tcp.h"), "File not found: /usr/include/netinet/tcp.h"
+    with open("/usr/include/netinet/tcp.h", "r") as fh:
```
This should be a stable user space ABI, I do not think you have to parse this header file.
Have you seen different versions of this structure?

jokujossai replied:
Changed to use `ctypes.Structure` implementation. [Musl](https://git.musl-libc.org/cgit/musl/tree/include/netinet/tcp.h#n195) have different tcp_info structure

hauke commented on scripts/flashing/dlink_recovery_upload.py:
```diff
@@ -0,0 +1,337 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: MIT
+#
+#   D-Link COVR-P2500 flash utility
```
Please explain what this utility is doing.
It looks like you try to not send more than one TCP packet at a time. 
Is this not possible by manipulating the TCP window size?

jokujossai replied:
Added explanation: "Upload firmware to the device with waiting for ACK after each TCP packet."
I did not manage to get the upload to work by manipulating the TCP window size or other socket settings.

s-2 commented:
I remember when we locally discussed flashing devices with older bootloaders recently and I mentioned this script, @lemoer suggested setting the `SO_SNDBUF` parameter to 0, disabling any buffer in the network stack and effectively making the calls block until the packet is sent. However I have not yet found the time to try and implement this, to see if flashing would work just as reliably as with polling the ACK status.

As a side note, due to the cheaply-available COVR-X1860 devices, a patch was sent to the mailing list a couple days ago, that adds a tool `dlink-sge-image` to firmware-utils, which allows for creating factory images also for the COVR-P2500.
Regardless of that, we would still need a flashing script in some way or another, that will allow recovering bricked devices via the bootloader.

s-2 commented:
`dlink-sge-image` is in `firmware-utils` now (bump PR at https://github.com/openwrt/openwrt/pull/13861).

I'm not sure how we shall proceed from here, since adding the encrypted image somehow creates a circular dependency with adding support for COVR-C1200 :joy:

Initially, I tried to add support for COVR-C1200 in https://github.com/openwrt/openwrt/pull/4174, which had the dtsi already separated for later adding of COVR-P2500, since I had considered C1200 easier to review (P2500 is very similar, just with additional PLC support).

The script `dlink-sge-signature` is probably no longer needed, as it turns out the fancy gzip-based signature is just a creative way of generating a crc32, so everything that script does could happen in a buildstep (I believe I had already done this at some point, but can't find that commit anymore :innocent: maybe it was just in my head).

However, I would like to wait for this PR first before adding C1200, especially since we also need your upload script (to recover bricked devices, or revert to OEM firmware).
I you'd like to update this PR to add an encrypted factory image besides recovery, I can have a look at the build recipe for the "md5 + device model + crc32" signature (as far as I recall it was slightly different between C1200 and P2500, i.e. one of them needed this at the end of the encrypted image, the other after the unencrypted image or something).

jokujossai commented:
> `dlink-sge-image` is in `firmware-utils` now (bump PR at #13861).

I created [commit](https://github.com/jokujossai/openwrt/commit/4accdb2cf3768545635d6a22e7c3c70cd3abff3b) with factory image support, but I didn't have time to test changes, so I didn't update this PR yet.

> The script `dlink-sge-signature` is probably no longer needed, as it turns out the fancy gzip-based signature is just a creative way of generating a crc32, so everything that script does could happen in a buildstep (I believe I had already done this at some point, but can't find that commit anymore 😇 maybe it was just in my head).

I modified `dlink-sge-signature.py` to call `binascii.crc32` instead of gzip.

jokujossai commented:
Updated PR with encrypted factory image support and added missing `dlink,covr-p2500-a1` section to `target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata`

DragonBluep commented on scripts/dlink-sge-signature.py:
```diff
@@ -0,0 +1,35 @@
+
+    fd.write(m.hexdigest().encode())
+    fd.write(f"\n{model_name}\n".encode())
+    fd.write("{:08x}".format(crc32).encode())
```
It should not be difficult to re-write it to makefile format using some shell commands.

jokujossai replied:
Converted to shell commands

s-2 replied:
> Converted to shell commands

I just realized `gzip` is indeed a widely used way of generating a crc32 in OpenWrt image recipes... :see_no_evil:

DragonBluep commented on scripts/flashing/dlink_recovery_upload.py:
```diff
@@ -0,0 +1,382 @@
+#
+#   D-Link COVR-P2500 flash utility
+#
+#   Upload firmware to the device with waiting for ACK after each TCP packet.
```
Attaching this file to the wiki should be more useful than integrating it into the source file.

s-2 replied:
There are lots of scripts in that folder, e.g. for AVM FritzBoxes. I think it's quite handy to have them checked out automatically with OpenWrt, but for the average user they should of course be mentioned in the wiki as well.

jokujossai replied:
Attached script to [wiki](https://openwrt.org/inbox/toh/d-link/covr-p2500_a1#dlink_recovery_uploadpy) and removed from git

DragonBluep commented on target/linux/ath79/dts/qca9563_dlink_covr-p2500-a1.dts:
```diff
@@ -0,0 +1,179 @@
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0>;
```
I remember this property has been removed from the ath driver.

DragonBluep replied:
Ref: https://github.com/openwrt/openwrt/commit/dae392752761128e13ede484217d806d643cbf2f

jokujossai replied:
Fixed

DragonBluep commented on target/linux/ath79/dts/qca9563_dlink_covr-p2500-a1.dts:
```diff
@@ -0,0 +1,179 @@
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		qca,mib-poll-interval = <500>;
+		reset-gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
```
The local ath driver doesn't support reset gpio. Please use `gpio-hog` to pull it high forever.

s-2 replied:
wait, is this a regression introduced with DSA migration here? (does ath79 use DSA fpr the gigabit switches at all?)
I had introduced the `reset-gpios` declaration for DIR-842 and DIR-859, and it definitely worked for those devices a couple years ago (i.e. changing it to `GPIO_ACTIVE_HIGH` was confirmed to have the switch disabled)

DragonBluep replied:
> wait, is this a regression introduced with DSA migration here? (does ath79 use DSA / `qca8k` for the gigabit switches at all?)

Currently, it is still a legacy swconfig driver. In
https://github.com/openwrt/openwrt/tree/main/target/linux/generic/files/drivers/net/phy


> I had introduced the `reset-gpios` declaration for DIR-842 and DIR-859, and it definitely worked for those devices a couple years ago (i.e. changing it to `GPIO_ACTIVE_HIGH` was confirmed to have the switch disabled)

No, that's because reset GPIO did not take effect at all. See: https://github.com/openwrt/openwrt/pull/12997#issuecomment-1612444072

jokujossai replied:
Replaced `reset-gpios` with
```
&gpio {
	phy-reset {
		gpio-hog;
		gpios = <11 GPIO_ACTIVE_HIGH>;
		output-high;
		line-name = "phy-reset";
	};
};
```

DragonBluep replied:
`GPIO_ACTIVE_LOW` and `output-low` should be better. Low output means that phy should not be reset.

jokujossai replied:
Fixed
```
&gpio {
	phy-reset {
		gpio-hog;
		gpios = <11 GPIO_ACTIVE_LOW>;
		output-low;
		line-name = "phy-reset";
	};
};
```

DragonBluep commented on target/linux/ath79/dts/qca9563_dlink_covr-p2500-a1.dts:
```diff
@@ -0,0 +1,179 @@
+&wmac {
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
```
Please convert it into NVMEM format. You can find a lot of examples on ath79 target.

jokujossai replied:
Converted to NVMEM and also converted `mac-address` to NVMEM, but I am not sure whether offset is same on every device. Previous implementation retrieved the value with `mtd_get_mac_ascii art "protest_ath0_mac"` and on my devices offset for this value is 0x95

jokujossai replied:
Changed `mac-address` logic back to `mtd_get_mac_ascii`

DragonBluep commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -196,6 +196,13 @@ case "$FIRMWARE" in
+	dlink,covr-p2500-a1)
+		caldata_extract "art" 0x5000 0x2f20
+		ath10k_patch_mac $(mtd_get_mac_ascii art "protest_ath1_mac")
+		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
```
I think this part can also be converted to `pre-calibration` NVMEM, too.

DragonBluep replied:
Ref: https://git.openwrt.org/openwrt/openwrt.git?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=grep&s=pre-calibration

jokujossai replied:
Converted `pre-calibration` and `mac-address` to NVMEM. I am not sure whether offset for `mtd_get_mac_ascii art "protest_ath1_mac"` is same `0xb8` on every device
```
00000000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000020  63 68 65 63 6b 73 75 6d  3d 58 58 58 58 58 58 58  |checksum=XXXXXXX|
00000030  58 0a 66 61 63 74 6f 72  79 6d 6f 64 65 3d 30 0a  |X.factorymode=0.|
00000040  70 72 6f 74 65 73 74 5f  70 6c 63 5f 6d 61 63 3d  |protest_plc_mac=|
00000050  58 58 3a 58 58 3a 58 58  3a 58 58 3a 58 58 3a 30  |XX:XX:XX:XX:XX:0|
00000060  31 0a 70 72 6f 74 65 73  74 5f 6c 61 6e 5f 6d 61  |1.protest_lan_ma|
00000070  63 3d 58 58 3a 58 58 3a  58 58 3a 58 58 3a 58 58  |c=XX:XX:XX:XX:XX|
00000080  3a 30 32 0a 70 72 6f 74  65 73 74 5f 61 74 68 30  |:02.protest_ath0|
00000090  5f 6d 61 63 3d 58 58 3a  58 58 3a 58 58 3a 58 58  |_mac=XX:XX:XX:XX|
000000a0  3a 58 58 3a 30 33 0a 70  72 6f 74 65 73 74 5f 61  |:XX:03.protest_a|
000000b0  74 68 31 5f 6d 61 63 3d  58 58 3a 58 58 3a 58 58  |th1_mac=XX:XX:XX|
000000c0  3a 58 58 3a 58 58 3a 30  34 0a 70 6c 63 5f 64 65  |:XX:XX:04.plc_de|
000000d0  6b 3d 58 58 58 58 2d 58  58 58 58 2d 58 58 58 58  |k=XXXX-XXXX-XXXX|
000000e0  2d 58 58 58 58 0a 70 6c  63 5f 6e 65 74 77 6f 72  |-XXXX.plc_networ|
000000f0  6b 70 77 64 3d 58 58 58  58 2d 58 58 58 58 2d 58  |kpwd=XXXX-XXXX-X|
00000100  58 58 58 2d 58 58 58 58  0a 73 73 69 64 3d 64 6c  |XXX-XXXX.ssid=dl|
00000110  69 6e 6b 2d 58 58 58 58  0a 77 6c 61 6e 70 61 73  |ink-XXXX.wlanpas|
00000120  73 77 64 3d 68 65 6d 66  6c 58 58 58 58 58 0a 70  |swd=hemflXXXXX.p|
00000130  6c 63 5f 63 6f 75 6e 74  72 79 63 6f 64 65 3d 47  |lc_countrycode=G|
```

DragonBluep replied:
> Converted `pre-calibration` and `mac-address` to NVMEM. I am not sure whether offset for `mtd_get_mac_ascii art "protest_ath1_mac"` is same `0xb8` on every device

These MAC addresses are generated and written by the same program, so I think they should have the same offset.

jokujossai replied:
Changed `mac-address` logic back to `mtd_get_mac_ascii`

s-2 commented:
Looks good :slightly_smiling_face: 

I just checked my old branch for COVR-C1200 to see what exactly was different between both models,
and back then it was only the LEDs that were individual, with everything else coming from a dtsi shared with P2500:
https://github.com/s-2/openwrt/tree/covr-c1200_master

I wonder what would be the most efficient way of adding C1200 later, but I think now it's probably easier to split the dts later in a separate PR for that device? Since the introduction of `nvmem-cells` for mac addresses and caldata, it seems extra care is needed regarding the hardcoded offsets of MAC addresses (which were parsed by a userspace script from the equals-separated config values), I'll have to check various versions of C1200 for the values to be consistent.

jokujossai commented:
> I wonder what would be the most efficient way of adding C1200 later, but I think now it's probably easier to split the dts later in a separate PR for that device? Since the introduction of `nvmem-cells` for mac addresses and caldata, it seems extra care is needed regarding the hardcoded offsets of MAC addresses (which were parsed by a userspace script from the equals-separated config values), I'll have to check various versions of C1200 for the values to be consistent.

I created separate [qca9563_dlink_covr](https://github.com/jokujossai/openwrt/tree/qca9563_dlink_covr) branch with splitted dts and C1200 support. I don't have C1200 to fetch the nvmem offsets or test on the device.

s-2 commented:
> I created separate [qca9563_dlink_covr](https://github.com/jokujossai/openwrt/tree/qca9563_dlink_covr) branch with splitted dts and C1200 support. I don't have C1200 to fetch the nvmem offsets or test on the device.

Thanks, I will dig out a few revisions of C1200 over the weekend and verify the offsets 🙂

s-2 commented:
I sifted through my piles of routers and selected 4 different versions of COVR-C1200:

* A1E with Bootloader v1.01, sold in Europe
* A1E with Bootloader v1.02, sold in Australia
* A1 with Bootloader v1.02, sold in Canada
* A1E with Bootloader v1.03, sold in Europe

At least the first three indeed share the same `protest_lan_mac` offsets of `0x65`, however the European version shipped with bootloader `1.03` (as per the label, not sure what the actual criteria are) has an additional field `fota_client_id` which breaks the mac address offsets.

<details>
<summary>ART dumps of various C1200</summary>

```
$ hexdump -C a1e_eu_1.01.bin | head -8
00000000  18 0f 76 b3 d4 68 ff ff  ff ff ff ff ff ff ff ff  |..v..h..........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  63 68 65 63 6b 73 75 6d  3d 31 35 33 36 65 38 39  |checksum=1536e89|
00000030  66 0a 66 61 63 74 6f 72  79 6d 6f 64 65 3d 30 0a  |f.factorymode=0.|
00000040  62 61 72 63 6f 64 65 3d  32 31 38 32 38 31 38 30  |barcode=21828180|
00000050  30 37 30 35 0a 70 72 6f  74 65 73 74 5f 6c 61 6e  |0705.protest_lan|
00000060  5f 6d 61 63 3d 31 38 3a  30 46 3a 37 36 3a 42 33  |_mac=18:0F:76:B3|
00000070  3a 44 34 3a 36 38 0a 64  65 76 69 63 65 5f 70 69  |:D4:68.device_pi|

$ hexdump -C a1e_au_1.02.bin | head -8
00000000  0c b6 d2 31 ab 40 ff ff  ff ff ff ff ff ff ff ff  |...1.@..........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  63 68 65 63 6b 73 75 6d  3d 31 35 33 36 65 38 39  |checksum=1536e89|
00000030  66 0a 66 61 63 74 6f 72  79 6d 6f 64 65 3d 30 0a  |f.factorymode=0.|
00000040  62 61 72 63 6f 64 65 3d  32 31 38 33 31 38 38 30  |barcode=21831880|
00000050  30 31 33 46 0a 70 72 6f  74 65 73 74 5f 6c 61 6e  |013F.protest_lan|
00000060  5f 6d 61 63 3d 30 43 3a  42 36 3a 44 32 3a 33 31  |_mac=0C:B6:D2:31|
00000070  3a 41 42 3a 34 30 0a 64  65 76 69 63 65 5f 70 69  |:AB:40.device_pi|

$ hexdump -C a1_ca_1.02.bin | head -8
00000000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000020  63 68 65 63 6b 73 75 6d  3d 63 65 33 64 37 33 31  |checksum=ce3d731|
00000030  63 0a 66 61 63 74 6f 72  79 6d 6f 64 65 3d 30 0a  |c.factorymode=0.|
00000040  62 61 72 63 6f 64 65 3d  32 31 38 33 39 38 34 30  |barcode=21839840|
00000050  30 30 30 45 0a 70 72 6f  74 65 73 74 5f 6c 61 6e  |000E.protest_lan|
00000060  5f 6d 61 63 3d 30 43 3a  42 36 3a 44 32 3a 35 31  |_mac=0C:B6:D2:51|
00000070  3a 32 36 3a 41 32 0a 64  65 76 69 63 65 5f 70 69  |:26:A2.device_pi|

$ hexdump -C a1e_eu_1.03.bin | head -24
00000000  0c b6 d2 5a 5e 4c ff ff  ff ff ff ff ff ff ff ff  |...Z^L..........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  63 68 65 63 6b 73 75 6d  3d 33 66 33 32 37 61 36  |checksum=3f327a6|
00000030  35 0a 66 61 63 74 6f 72  79 6d 6f 64 65 3d 30 0a  |5.factorymode=0.|
00000040  66 6f 74 61 5f 63 6c 69  65 6e 74 5f 69 64 3d 30  |fota_client_id=0|
00000050  66 31 62 35 35 34 39 62  64 38 66 65 66 65 61 35  |f1b5549bd8fefea5|
00000060  34 36 63 61 64 38 33 31  38 64 33 35 30 61 61 62  |46cad8318d350aab|
00000070  37 61 66 61 38 65 39 32  33 64 39 64 35 36 65 62  |7afa8e923d9d56eb|
00000080  37 34 63 31 63 62 32 36  30 30 37 65 30 61 65 37  |74c1cb26007e0ae7|
00000090  61 30 33 63 34 38 32 61  32 39 35 39 30 31 62 62  |a03c482a295901bb|
000000a0  65 65 64 64 61 31 31 33  36 30 39 33 39 65 61 0a  |eedda11360939ea.|
000000b0  66 6f 74 61 5f 63 6c 69  65 6e 74 5f 73 65 63 72  |fota_client_secr|
000000c0  65 74 3d 35 63 31 62 38  63 35 65 66 36 37 33 61  |et=5c1b8c5ef673a|
000000d0  35 62 37 36 39 32 36 32  38 37 63 64 64 32 35 37  |5b76926287cdd257|
000000e0  30 37 37 35 62 32 30 61  61 66 35 61 61 63 39 32  |0775b20aaf5aac92|
000000f0  65 36 66 32 65 35 37 31  31 38 36 64 32 38 32 61  |e6f2e571186d282a|
00000100  32 63 34 37 61 30 33 63  34 38 32 61 32 39 35 39  |2c47a03c482a2959|
00000110  30 31 62 62 65 65 64 64  61 31 31 33 36 30 39 33  |01bbeedda1136093|
00000120  39 65 61 0a 62 61 72 63  6f 64 65 3d 32 31 38 34  |9ea.barcode=2184|
00000130  31 32 36 30 30 31 46 42  0a 70 72 6f 74 65 73 74  |126001FB.protest|
00000140  5f 6c 61 6e 5f 6d 61 63  3d 30 43 3a 42 36 3a 44  |_lan_mac=0C:B6:D|
00000150  32 3a 35 41 3a 35 45 3a  34 43 0a 64 65 76 69 63  |2:5A:5E:4C.devic|
00000160  65 5f 70 69 6e 3d 35 39  32 32 33 38 30 31 0a 70  |e_pin=59223801.p|
00000170  72 6f 74 65 73 74 5f 77  61 6e 5f 6d 61 63 3d 30  |rotest_wan_mac=0|
```

</details>

So I believe there is no clean way to hardcode the offsets (like providing a different set of images for each variant etc.),
as there might be even more variants, depending on how the config was generated. 

For COVR-P2500 it might be okay though, considering there only seems to be the EU version from what I've seen (due to the power plug being part of the case?), and my device has it consistent with yours at `0x72` (so it's yet different from any of the C1200 and could not share a .dtsi here as well). 

As far as I recall, the [parsing vs. hardcoding](https://github.com/openwrt/openwrt/pull/3997#discussion_r1071632465) (in favor of being upstream-compatible) has previously been a discussion with other devices as well.
fun fact: at least for caldata, there even has been some [trial and error magic](https://github.com/openwrt/openwrt/blob/935a63c59d634ee384635003391b65399097f683/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata#L203) implemented due to some revisions of DIR-842, to find the correct location of caldata...

jokujossai commented:
> So I believe there is no clean way to hardcode the offsets (like providing a different set of images for each variant etc.), as there might be even more variants, depending on how the config was generated.
> 
> For COVR-P2500 it might be okay though, considering there only seems to be the EU version from what I've seen (due to the power plug being part of the case?), and my device has it consistent with yours at `0x72` (so it's yet different from any of the C1200 and could not share a .dtsi here as well).

Changed MAC address logics back to `mtd_get_mac_ascii`. COVR-P2500 does not have value for `protest_wan_mac` so I configured both lan and wan to use same `protest_lan_mac` value. Also configured `protest_plc_mac` for the PLC interface

s-2 commented:
> Changed MAC address logics back to `mtd_get_mac_ascii`.

Thanks, I hope this does not get in the way of a later migration to DSA, but the worst case scenario would be to maintain a patch that implements this parsing downstream, way better than risking to have devices with weird MACs out there.

I'm not sure if I confused it with the `mac_text` format, but I believe I've seen such an attempt before, could not find it anywhere though.

hauke commented on include/image-commands.mk:
```diff
@@ -226,6 +226,11 @@ define Build/copy-file
 endef
 
+define Build/dlink-sge-image
+	$(STAGING_DIR_HOST)/bin/dlink-sge-image $(1) $@ $@.enc
```
This was merged into master in 4c0dc68f466, please remove it here.

jokujossai replied:
fixed

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -995,6 +1006,28 @@ define Device/devolo_magic-2-wifi
+  DEVICE_VARIANT := A1
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca9888-ct \
+	open-plc-utils open-plc-utils-hpavkey open-plc-utils-modpib \
+	open-plc-utils-plchost open-plc-utils-plctool
```
The open-plc applications are not part of the main openwrt repository, but the feed, please remove them here.

jokujossai replied:
Fixed

hauke commented:
Thank you for the PR, I merged it into main in a39a49e323e6731215bd2654e68d91a2df4672c7.
